plugins/bcm57xx/fu-bcm57xx-dict-image.c

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (C) 2020 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-common.h"
+
+#include "fu-bcm57xx-common.h"
+#include "fu-bcm57xx-dict-image.h"
+
+struct _FuBcm57xxDictImage {
+	FuFirmwareImage		 parent_instance;
+	guint8			 target;
+	guint8			 kind;
+};
+
+G_DEFINE_TYPE (FuBcm57xxDictImage, fu_bcm57xx_dict_image, FU_TYPE_FIRMWARE_IMAGE)
+
+static void
+fu_bcm57xx_dict_image_to_string (FuFirmwareImage *image, guint idt, GString *str)
+{
+	FuBcm57xxDictImage *self = FU_BCM57XX_DICT_IMAGE (image);
+	if (self->target != 0xff)
+		fu_common_string_append_kx (str, idt, "Target", self->target);
+	if (self->kind != 0xff)
+		fu_common_string_append_kx (str, idt, "Kind", self->kind);
+}
+
+static gboolean
+fu_bcm57xx_dict_image_parse (FuFirmwareImage *image,
+			     GBytes *fw,
+			     FwupdInstallFlags flags,
+			     GError **error)
+{
+	g_autoptr(GBytes) fw_nocrc = NULL;
+	if ((flags & FWUPD_INSTALL_FLAG_FORCE) == 0) {
+		if (!fu_bcm57xx_verify_crc (fw, error))
+			return FALSE;
+	}
+	fw_nocrc = g_bytes_new_from_bytes (fw, 0x0, g_bytes_get_size (fw) - sizeof(guint32));
+	fu_firmware_image_set_bytes (image, fw_nocrc);
+	return TRUE;
+}
+
+static GBytes *
+fu_bcm57xx_dict_image_write (FuFirmwareImage *image, GError **error)
+{
+	const guint8 *buf;
+	gsize bufsz = 0;
+	guint32 crc;
+	g_autoptr(GByteArray) blob = NULL;
+	g_autoptr(GBytes) fw_nocrc = NULL;
+
+	/* get the CRC-less data */
+	fw_nocrc = fu_firmware_image_get_bytes (image);
+	if (fw_nocrc == NULL || g_bytes_get_size (fw_nocrc) == 0)
+		return g_bytes_new (NULL, 0);
```

Comment:
don't this need to set error?  Or it's expected scenario?

Reply:
That's a good question. `fw_nocrc == NULL` should be the error, and even for an empty section we should write a CRC.

Fixed.